### ISSUE TITLE
Coercion of large HDF5Matrix to SparseArraySeed fails

### ISSUE THREAD

**Issue by LTLA** [Bioconductor/HDF5Array#27]
I have a moderately large HDF5Matrix (50k by 50k) and its coercion fails with:

```r
Y <- as(X, "SparseArraySeed")
## Error: subscript contains NAs
## In addition: Warning messages:
## 1: In (ans - 1L) * dim[, along] : NAs produced by integer overflow
## 2: In (ans - 1L) * dim[, along] + aind[, along] :
##   NAs produced by integer overflow
traceback()
## 10: stop(wmsg(...), call. = FALSE)
## 9: .subscript_error("subscript contains NAs")
## 8: NSBS(i, x, exact = exact, strict.upper.bound = !allow.append,
##        allow.NAs = allow.NAs)
## 7: NSBS(i, x, exact = exact, strict.upper.bound = !allow.append,
##        allow.NAs = allow.NAs)
## 6: normalizeSingleBracketSubscript(i, x)
## 5: normalizeSingleBracketSubscript2(i, length(x))
## 4: from[idx]
## 3: from[idx]
## 2: asMethod(object)
## 1: as(counts(se), "SparseArraySeed")
```

Some digging indicates that it fails at the `nzdata <- from[idx]` line in the `coerce` method for `HDF5Matrix` to `SparseArraySeed`, with the warnings generated by code in `BLOCK_which`. I presume that this because it is trying to extract a value at the bottom-right of the matrix by treating it as a long vector, in which case the final index exceeds `.Machine$integer.max`.

<details>
<summary>Session info</summary>

```
R version 3.6.1 (2019-07-05)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS:   /gstore/apps/R/R_3.6.1_Bioc_3.10/R-3.6.1-Bioc-3.10-uat-20191108/lib64/R/lib/libRblas.so
LAPACK: /gstore/apps/R/R_3.6.1_Bioc_3.10/R-3.6.1-Bioc-3.10-uat-20191108/lib64/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] parallel  stats4    stats     graphics  grDevices utils     datasets
[8] methods   base

other attached packages:
 [1] HDF5Array_1.14.2            rhdf5_2.30.1
 [3] SingleCellExperiment_1.8.0  gp.sa.io_1.0.29
 [5] SummarizedExperiment_1.16.1 DelayedArray_0.12.2
 [7] BiocParallel_1.20.1         matrixStats_0.55.0
 [9] Biobase_2.46.0              GenomicRanges_1.38.0
[11] GenomeInfoDb_1.22.0         IRanges_2.20.2
[13] S4Vectors_0.24.3            BiocGenerics_0.32.0

loaded via a namespace (and not attached):
 [1] tidyselect_1.0.0            xfun_0.12
 [3] purrr_0.3.3                 lattice_0.20-40
 [5] vctrs_0.2.3                 htmltools_0.4.0
 [7] BiocFileCache_1.10.2        yaml_2.2.1
 [9] blob_1.2.1                  rlang_0.4.4
[11] R.oo_1.23.0                 pillar_1.4.3
[13] R.utils_2.9.2               glue_1.3.1
[15] DBI_1.1.0                   rappdirs_0.3.1
[17] bit64_0.9-7                 dbplyr_1.4.2
[19] GenomeInfoDbData_1.2.2      zlibbioc_1.32.0
[21] R.methodsS3_1.8.0           memoise_1.1.0
[23] evaluate_0.14               knitr_1.28
[25] callr_3.4.2                 ps_1.3.2
[27] MultiAssayExperiment_1.12.2 curl_4.3
[29] Rcpp_1.0.3                  BiocManager_1.30.10
[31] XVector_0.26.0              bit_1.1-15.2
[33] BiocStyle_2.14.4            digest_0.6.25
[35] processx_3.4.2              dplyr_0.8.4
[37] getPass_0.2-2               grid_3.6.1
[39] tools_3.6.1                 bitops_1.0-6
[41] magrittr_1.5                RCurl_1.98-1.1
[43] RSQLite_2.2.0               tibble_2.1.3
[45] crayon_1.3.4                pkgconfig_2.0.3
[47] Matrix_1.2-18               assertthat_0.2.1
[49] rmarkdown_2.1               httr_1.4.1
[51] Rhdf5lib_1.8.0              gp.sa.core_1.0.18
[53] R6_2.4.1                    compiler_3.6.1
```
</details>

**Comment by hpages** [comment]
mmh I see, at the very low level is `linearInd()` not being able to compute linear indices greater that `.Machine$integer.max`:
```
aind <- cbind(c(1, 2, 49999, 50000), c(1, 1, 50000, 50000))
aind
#       [,1]  [,2]
# [1,]     1     1
# [2,]     2     1
# [3,] 49999 50000
# [4,] 50000 50000

linearInd(aind, c(50000L, 50000L))
# [1]  1  2 NA NA
# Warning message:
# In (ans - 1L) * dim[, along] : NAs produced by integer overflow
```
This is implemented in R at the moment and the fix is easy. However it's one of those things that is probably worth moving to the C level so I think I'm gonna do that instead.

Thanks,
H.

**Comment by hpages** [comment]
Should be fixed in DelayedArray 0.13.10. Bonus: coercion from HDF5Matrix to SparseArraySeed or dgCMatrix got a 2x speedup.

**Comment by LTLA** [comment]
Very nice, thank you.

**Comment by LTLA** [comment]
Could this fix get propagated to release as well?

**Comment by hpages** [comment]
I guess I could. There are many changes involved (basically almost a complete refactor of the 1D-style single bracket subsetting code) so I'll wait 24h for it to propagate to devel, and, if everything looks ok, I'll port to release.

**Comment by hpages** [comment]
The fix is now in release too (in DelayedArray 0.12.3).

**Comment by LTLA** [comment]
Thanks, this will help a lot.
